proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+
+```go
+// NestedControlPlaneTemplateSpec defines the desired state of NestedControlPlaneTemplate
+type NestedControlPlaneTemplateSpec struct {
+    // other fields ...    
+    
+    // NestedEtcd contains information required by the etcd provider to create the Etcd
+    NestedEtcd NestedComponentSpec `json:"nestedEtcd"`
+
+    // NestedKAS contains information required by the apiserver provider to create the apiserver
+    NestedKAS NestedComponentSpec `json:"nestedKAS"`
+
+    // NestedKCM contains information required by the controller-manager provider to create the controller-manager
+    NestedKCM NestedComponentSpec `json:"nestedKCM"`
+}
+```
+
+We will let users specify component providers through NCP, implement a native mode that has each sub-controller to create the component on the super cluster, and define three new types `EtcdProvider`, `KASProvider`, and the `KCMProvider`:
+
+```go
+type KASProvider string
+
+type EtcdProvider string
+
+type KCMProvider string
+
+const (
+    NativeKASProvider KASProvider = "NativeKASProvider"
+
+    NativeEtcdProvider  EtcdProvider = "NativeEtcdProvider"
+    EtcdClusterOperator EtcdProvider = "EtcdClusterOperator"
+    EtcdOperator        EtcdProvider = "EtcdOperator"
+
+    NativeKCMProvider KCMProvider = "NativeKCMProvider"
+)
+```
+
+We will add three new fields `NestedEtcdProvider`, `NestedKASProvider`, and the `NestedKCMProvider` into the NCP, which will make the NCP look like the following:
+
+```go
+// NestedControlPlaneSpec defines the desired state of NestedControlPlane
+type NestedControlPlaneSpec struct {
+    // other fields.... 
+    
+    // NestedEtcdProvider specifies which EtcdProvider will be used to create the Etcd
+    // +optional
+    NestedEtcdProvider EtcdProvider `json:"etcdProvider,omitempty"`
+     
+    // NestedKASProvider specifies which KASProvider will be used to create the kube-apiserver
+    // +optional
+    NestedKASProvider KASProvider `json:"kasProvider,omitempty"`
+ 
+    // NestedKCMProvider specifies which KCMProvider will be used to create the kube-controller-manager
+    // +optional
+    NestedKCMProvider KCMProvider `json:"kcmProvider,omitempty"`
+}
+```
+
+### Bootstrap
+
+#### Create prerequisites 
+
+The NCP controller will need to create some resources before sub-controllers start creating the real NCP components.
+
+* **ClusterNamespace** is used to store all resources for the NCP.
+
+* **NKAS service**  is used to visit the NKAS. Since the IP/domain name is required when generating TLS certificates and kubeconfigs of NKAS. This service will need to be created in advance.
+
+* **NestedCRs**, including NEtcd, NKAS, and NKCM. are CRs that hold information required by the sub-controller. 
```

Comment:
Let's start without NCP creating the nested CRs since we won't have templates, so to create a nested cluster you need to create each of NCluster, NCP, NEtcd, NKAS, NKCM and these will work together to create the control plane.